repository-gcs/src/main/java/io.aiven.elasticsearch.repositories.gcs/GcsSettingsProvider.java

```diff
@@ -18,69 +18,46 @@
+import io.aiven.elasticsearch.repositories.RepositoryStorageIOProvider;
 import io.aiven.elasticsearch.repositories.security.EncryptionKeyProvider;
 
-import com.google.cloud.Tuple;
 import com.google.cloud.http.HttpTransportOptions;
 import com.google.cloud.storage.Storage;
 import com.google.cloud.storage.StorageOptions;
 import com.google.common.base.Strings;
 import com.google.common.net.HttpHeaders;
 import org.elasticsearch.common.settings.Settings;
 
-public class GcsSettingsProvider {
+public class GcsSettingsProvider extends RepositorySettingsProvider<Storage> {
 
-    static final String HTTP_USER_AGENT = "Aiven Google Cloud Storage repository";
+    static final String HTTP_USER_AGENT = "Aiven GCS Repository";
 
-    private volatile Tuple<Storage, EncryptionKeyProvider> cachedSettings;
-
-    public Storage gcsClient() throws IOException {
-        //we should throw IOException for such action. ES swallows others
-        if (Objects.isNull(cachedSettings) || Objects.isNull(cachedSettings.x())) {
-            throw new IOException("GCS client hasn't been configured");
-        }
-        return cachedSettings.x();
+    @Override
+    protected RepositoryStorageIOProvider<Storage> createRepositoryStorageIOProvider(
+            final Settings settings, final EncryptionKeyProvider encryptionKeyProvider) throws IOException {
+        final var client = Permissions.doPrivileged(() -> createGcsClient(GcsStorageSettings.create(settings)));
+        return new GcsRepositoryStorageIOProvider(client, encryptionKeyProvider);
     }
 
-    public EncryptionKeyProvider encryptionKeyProvider() throws IOException {
-        //we should throw IOException for such action. ES swallows others
-        if (Objects.isNull(cachedSettings) || Objects.isNull(cachedSettings.y())) {
-            throw new IOException("EncryptionKeyProvider hasn't been configured");
+    private Storage createGcsClient(final GcsStorageSettings gcsStorageSettings) throws IOException {
+        final StorageOptions.Builder storageOptionsBuilder = StorageOptions.newBuilder();
+        if (!Strings.isNullOrEmpty(gcsStorageSettings.projectId())) {
+            storageOptionsBuilder.setProjectId(gcsStorageSettings.projectId());
         }
-        return cachedSettings.y();
-    }
-
-    public void reload(final Settings settings) throws IOException {
-        cachedSettings = makeCacheFrom(settings);
+        storageOptionsBuilder
+                .setTransportOptions(
+                        HttpTransportOptions.newBuilder()
+                                .setConnectTimeout(gcsStorageSettings.connectionTimeout())
+                                .setReadTimeout(gcsStorageSettings.readTimeout())
+                                .build())
+                .setHeaderProvider(() -> Map.of(HttpHeaders.USER_AGENT, HTTP_USER_AGENT))
+                .setCredentials(gcsStorageSettings.gcsCredentials());
+
+        return storageOptionsBuilder.build().getService();
```

Comment:
This used to be inside `Permissions.doPrivileged`. Is it not needed any more?

Reply:
it is in the line #41 ... i think it is more readable